File: String to Integer (atoi).java
public class Solution {
    public int atoi(String str) {
        int status = 0;
        int sign = +1;
        long ret = 0;
        for (int i = 0; i < str.length(); i++) {
            switch (status) {
                case 0:
                    if (Character.isWhitespace(str.charAt(i)))
                        break;
                    else if (str.charAt(i) == '-') {
                        sign = -1;
                        status = 1;
                        break;
                    }
                    else if (str.charAt(i) == '+') {
                        sign = +1;
                        status = 1;
                        break;
                    }
                    status = 1;
                    // not break intentionally
                case 1:
                    if (Character.isDigit(str.charAt(i))) {
                        ret *= 10;
                        ret += str.charAt(i) - '0';
                        if (sign * ret > Integer.MAX_VALUE)
                            return Integer.MAX_VALUE;
                        if (sign * ret < Integer.MIN_VALUE)
                            return Integer.MIN_VALUE;
                    }
                    else
                        return (int)(sign * ret);
                    break;
            }
        }
        return (int)(sign * ret);
    }
}

File: String to Integer (atoi).py
class Solution:
	# @return an integer
	def fit_in_range(self, x):
		if (x > 2147483647):
			return 2147483647
		elif (x < -2147483648):
			return -2147483648
		else:
			return x
	def atoi(self, str):
		ret = 0
		sign = +1
		status = 0  # not number; might be sign symbol or white space
		for ch in str:
			if (status == 0):
				if (ch == ' ' or ch == '\t'):
					continue
				if (ch == '+'):
					status = 1
					continue
				if (ch == '-'):
					status = 1
					sign = -1
					continue
				if (ch >= '0' and ch <= '9'):
					status = 1
				else:
					return 0
			if (status == 1):
				if (ch >= '0' and ch <= '9'):
					ret = ret * 10 + ord(ch) - ord('0')
				else:
					return self.fit_in_range(sign * ret)
		return self.fit_in_range(sign * ret)
		